### PR TITLE
 0-RTT can't use transport parameters or 1-RTT frames 

### LINKED ISSUE #2360: 0-RTT flow control limits can't be increased
In #2344, we discuss what can be sent in 0-RTT.  We realized that while in theory the client should not be accepting 1-RTT packets from the server and then sending 0-RTT, there is nothing to prevent that.

The proposed defense is to prohibit the client from using increased limits (on flow control or streams) that it might have learned from 1-RTT packets.  This ensures that 0-RTT - the parts of the connection that might be replayed - are strictly limited to the initial limits.  Like many other things, we would require the client not to exceed this limits and merely allow the server to enforce that, in case the server was not inclined to do so.

### LINKED ISSUE #2458: Client MUST use 1-RTT packets if it reads 1-RTT packets
I am not sure if this is accurate.

IIUC their is nothing that mandates the client to switch to using 1-RTT when it receives ServerFinished. In fact, on a lossy network, a client would benefit from using 0-RTT packets until it receives an ACK for ClientFinished. In such case, it would be natural to observe these frames in a 0-RTT packet.

I think we should either forbid them or clarify that there are allowed, instead of suggesting (IMO incorrectly) that it’s impossible.

_Originally posted by @kazuho in https://github.com/_render_node/MDExOlB1bGxSZXF1ZXN0MjQ2MTcyNjUx/pull_requests/unread_timeline_

### PR DESCRIPTION
**Pull request by martinthomson**
While in theory a client could send 0-RTT after receiving 1-RTT, this is
not a good idea.  This PR ties down transport parameters and flow
control and prevents a client from using anything it learns in the
handshake or 1-RTT from affecting (or expanding) what it can do with
0-RTT.  This prevents a client from using 0-RTT indefinitely and creates
stronger bounds on what 0-RTT can be used for.

Closes #2458.
Closes #2360.
Closes #2466.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -3652,27 +3662,31 @@ resend data in 0-RTT packets after it sends a new Initial packet.
...
 
 A client SHOULD instead generate a fresh cryptographic handshake message and
 start packet numbers from 0.  This ensures that new 0-RTT packets will not use
 the same keys, avoiding any risk of key and nonce reuse; this also prevents
 0-RTT packets from previous handshake attempts from being accepted as part of
 the connection.
 
+A client MUST NOT send 0-RTT packets once it starts processing 1-RTT packets
+from the server.  This means that 0-RTT packets can't contain any response to
+packets a server sends.  For instance, a client cannot send an ACK frame in a
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: Maybe it's worth clarifying that: "any response to _1-RTT_ packets"?

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -3652,27 +3662,32 @@ resend data in 0-RTT packets after it sends a new Initial packet.
...
+0.  Thus, 0-RTT packets could need to use a longer packet number encoding.
 
 A client SHOULD instead generate a fresh cryptographic handshake message and
 start packet numbers from 0.  This ensures that new 0-RTT packets will not use
 the same keys, avoiding any risk of key and nonce reuse; this also prevents
 0-RTT packets from previous handshake attempts from being accepted as part of
 the connection.
 
+A client MUST NOT send 0-RTT packets once it starts processing 1-RTT packets
+from the server.  This means that 0-RTT packets can't contain any response to
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]:
```suggestion
from the server.  This means that 0-RTT packets cannot contain any response to
```

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -3652,27 +3662,32 @@ resend data in 0-RTT packets after it sends a new Initial packet.
...
+frames from 1-RTT packets.  For instance, a client cannot send an ACK frame in a
+0-RTT packet, because that can only acknowledge a 1-RTT packet.  An
+acknowledgment for a 1-RTT packet MUST be carried in a 1-RTT packet.
+
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: A server MAY close a connection with the error PROTOCOL_VIOLATION if it is able to observe that a client sends 0-RTT packets based on any information received in 1-RTT packets, such as using increased flow control credits or stream identifiers.

**Comment by kazuho** [comment]
@ianswett 
> I think this is ok, but I think it's going to be difficult to get this correct in practice. I expect most implementations will migrate to 1-RTT keys as soon as possible and try to avoid writing special logic for this case, but I can't come up with a way to force clients to do that.

I think it's possible. I think that's what the PR is trying to suggest is that a server can:
* enforce limits based on TP to streams-related frames sent using 0-RTT packets (see #2461)
* refuse to handle any of the following frames in a 0-RTT packet: ACK, PATH_RESPONSE, RETIRE_CONNECTION_ID (see #2458)

**Comment by MikeBishop** [comment]
> * enforce limits based on TP to streams-related frames sent using 0-RTT packets

Actually, enforce limits based on the TP *from last time* which are embedded in the session ticket.  But yes -- the "correct" way of doing this is saying that there is a parallel set of flow control limits which never increase from the initial value, but which only apply during 0-RTT.  If the client always switches to 1-RTT ASAP, this will be the same as regular flow control provided that the client doesn't consume any higher initial flow control values in the transport parameters until the handshake completes.

**Comment by martinthomson** [comment]
I tend to agree with @ianswett that policing limits here will be annoying, even if it is theoretically possible.  Thankfully, we get most of the benefits if only a small number of implementations police limits.

**Comment by DavidSchinazi** [comment]
I know we agreed to this is Tokyo, but seeing the actual PR made me realize how complex and hard to enforce this is. Would it be possible to solve the issue in a simpler way by saying something along the lines of the following? `Once a server receives a valid 1-RTT packet with packet number n, any 0-RTT packet it receives with packet number greater than n MUST be ignored.` That way clients can't keep sending 0-RTT forever since they'll have to send a 1-RTT ACK to unblock the server-to-client flow control at some point.

**Comment by martinthomson** [comment]
> Once a server receives a valid 1-RTT packet with packet number n, any 0-RTT packet it receives with packet number greater than n MUST be ignored.

This is already the case.  It fails to solve any of the problems we are faced with though.

**Comment by MikeBishop** [comment]
@DavidSchinazi, if you want to change the decision at the interim, that's decidedly a question for the issue/list.

> > Once a server receives a valid 1-RTT packet with packet number n, any 0-RTT packet it receives with packet number greater than n MUST be ignored.
> 
> This is already the case. It fails to solve any of the problems we are faced with though.

@martinthomson, can you point to that restriction?  I agree that it's intended, but I can't find it.

If we have that, you have to ignore 0-RTT packets sent after 1-RTT has started.  If you withhold putting any increases in flow control into effect until you've seen that 1-RTT packet, you're done fixing the problem we discussed.  That might be a simpler way to describe (or at least to implement) this prohibition.

**Comment by martinthomson** [comment]
https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#rfc.section.4.9 says:

> https://quicwg.org/base-drafts/draft-ietf-quic-tls.html#rfc.section.4.9

We already have issues on file for not reading 1-RTT and then sending more 0-RTT.  They even have PRs.

**Comment by MikeBishop** [comment]
Ah, I was looking in -transport.  That prohibits the client from going back to 0-RTT, and prohibits the server from accepting 0-RTT starting 3RTO after the server first sends a 1-RTT packet, but doesn't appear to require the server to check that the client didn't reverse itself by dropping packets.

(This *is* the issue/PR about not allowing the client to stay in 1-RTT, isn't it?)

**Comment by martinthomson** [comment]
Gah, too many windows.  Sorry.

**Comment by DavidSchinazi** [comment]
@MikeBishop I wouldn't say I'm explicitly trying to change anything decided in Tokyo, I'm trying to understand why I'm failing to understand what problem this PR solves.

@martinthomson the text you pasted from -tls doesn't match what I suggested:

> Once a server receives a valid 1-RTT packet with packet number n, any 0-RTT packet it receives with packet number greater than n MUST be ignored.

So I don't think we already do that. Could you elaborate on what problem is solved by this PR that wouldn't be solved by my check above?

**Comment by MikeBishop** [comment]
(To be clear, @martinthomson pasted the URL again as the quote, and I expanded it to some of the relevant text.)  I think what he was meaning to point to was this:
> Though an endpoint might retain older keys, new data MUST be sent at the highest
currently-available encryption level.  Only ACK frames and retransmissions of
data in CRYPTO frames are sent at a previous encryption level.  These packets
MAY also include PADDING frames.

So the client is prohibited from doing that, but the server isn't currently obligated to check.  I've tried to create something along the lines you suggest in #2466 -- can you take a look?

**Comment by martinthomson** [comment]
Too little sleep.  I still think that this is better than #2466.  Comments why over there.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1422,6 +1422,16 @@ server's new preferred_address value in the handshake.
 A server MUST either reject 0-RTT data or abort a handshake if the implied
 values for transport parameters cannot be supported.
 
+A client MUST NOT use updated values it learns either from the server's updated
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: This is difficult to read. How about:
"When sending frames in 0-RTT packets, a client MUST only use remembered transport parameters; importantly, it MUST NOT use updated values that it learns from the server's updated transport parameters or from frames received in 1-RTT packets."

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -1422,6 +1422,16 @@ server's new preferred_address value in the handshake.
...
+transport parameters or frames carried in 1-RTT when sending frames in 0-RTT
+packets.  Updated values of transport parameters from the handshake only apply
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
packets.  Updated values of transport parameters from the handshake apply only
```

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -1422,6 +1422,16 @@ server's new preferred_address value in the handshake.
...
+to 1-RTT packets.  For instance, flow control limits from remembered transport
+parameters apply to all 0-RTT packets even if those values are increased by the
+handshake; increased limits from frames cannot be applied, because the client
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
handshake.  Increased flow control limits from frames cannot be applied, because the client
```

Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I might even suggest dropping the second half, after "increased limits from frames".

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -3652,27 +3662,35 @@ resend data in 0-RTT packets after it sends a new Initial packet.
...
 start packet numbers from 0.  This ensures that new 0-RTT packets will not use
 the same keys, avoiding any risk of key and nonce reuse; this also prevents
 0-RTT packets from previous handshake attempts from being accepted as part of
 the connection.
 
+A client MUST NOT send 0-RTT packets once it starts processing 1-RTT packets
+from the server.  This means that 0-RTT packets cannot contain any response to
+frames from 1-RTT packets.  For instance, a client cannot send an ACK frame in a
+0-RTT packet, because that can only acknowledge a 1-RTT packet.  An
+acknowledgment for a 1-RTT packet MUST be carried in a 1-RTT packet.  A server
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: New paragraph -- this last sentence is not related to the earlier discussion in this para.
```suggestion
acknowledgment for a 1-RTT packet MUST be carried in a 1-RTT packet. 

A server
```

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -1422,6 +1422,15 @@ server's new preferred_address value in the handshake.
 A server MUST either reject 0-RTT data or abort a handshake if the implied
 values for transport parameters cannot be supported.
 
+When sending frames in 0-RTT packets, a client MUST only use remembered
+transport parameters; importantly, it MUST NOT use updated values that it learns
+from the server's updated transport parameters or from frames received in 1-RTT
+packets.  Updated values of transport parameters from the handshake apply only
+to 1-RTT packets.  For instance, flow control limits from remembered transport
+parameters apply to all 0-RTT packets even if those values are increased by the
+handshake or by frames send in 1-RTT packets.  A server MAY treat use of updated
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: s/send/sent/?

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -3652,27 +3661,36 @@ resend data in 0-RTT packets after it sends a new Initial packet.
...
 0-RTT packets from previous handshake attempts from being accepted as part of
 the connection.
 
+A client MUST NOT send 0-RTT packets once it starts processing 1-RTT packets
+from the server.  This means that 0-RTT packets cannot contain any response to
+frames from 1-RTT packets.  For instance, a client cannot send an ACK frame in a
+0-RTT packet, because that can only acknowledge a 1-RTT packet.  An
+acknowledgment for a 1-RTT packet MUST be carried in a 1-RTT packet.
+
+A server MAY treat a violation of remembered limits as a connection error of an
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]:
```suggestion
A server SHOULD treat a violation of remembered limits as a connection error of an
```
Per previous comment, I believe the outcome from Tokyo was to recommend enforcement.  Subsequent discussion has concluded that thorough enforcement might be too heavy-weight, but the specific flow-control approach is simple enough and ought to still be recommended.
